src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/aggregates/AggregateRewriting.java

```diff
@@ -0,0 +1,142 @@
+	 * During preprocessing, all aggregate literals with two comparison operators are split into two aggregate literals with
+	 * only a "lower bound" term and operator. After literal splitting, operators are normalized, i.e. all "count" and "sum"
+	 * literals are rewritten to use either "<=" or "=" as comparison operator.
+	 * 
+	 * Rules containing {@link AggregateLiteral}s are registered in an {@link AggregateRewritingContext} during
+	 * preprocessing. After preprocessing, for each rule in the context, aggregate literals in the rule body are substituted
+	 * with normal literals of form "$id$_aggregate_result(...)". For each literal substituted this way, a set of rules
+	 * deriving the result literal is added that is semantically equivalent to the replaced aggregate literal.
+	 */
+	@Override
+	// Note: Ideally, all variables introduced by AggregateRewriting should be prefixed "_AGG" or something.
+	public InputProgram apply(InputProgram inputProgram) {
+		AggregateRewritingContext ctx = new AggregateRewritingContext();
+		List<BasicRule> outputRules = new ArrayList<>();
+		for (BasicRule inputRule : inputProgram.getRules()) {
+			// Split literals with two operators.
+			for (BasicRule splitRule : AggregateLiteralSplitting.split(inputRule)) {
+				// Normalize operators on aggregate literals after splitting.
+				BasicRule operatorNormalizedRule = AggregateOperatorNormalization.normalize(splitRule);
+				boolean hasAggregate = ctx.registerRule(operatorNormalizedRule);
+				// Only keep rules without aggregates. The ones with aggregates are registered in the context and taken care of later.
+				if (!hasAggregate) {
+					outputRules.add(operatorNormalizedRule);
+				}
+			}
+		}
+		// Substitute AggregateLiterals with generated result literals.
+		outputRules.addAll(rewriteRulesWithAggregates(ctx));
+		InputProgram.Builder resultBuilder = InputProgram.builder().addRules(outputRules).addFacts(inputProgram.getFacts())
+				.addInlineDirectives(inputProgram.getInlineDirectives());
+		// Add sub-programs deriving respective aggregate literals.
+		for (ImmutablePair<AggregateFunctionSymbol, ComparisonOperator> func : ctx.getAggregateFunctionsToRewrite().keySet()) {
+			AbstractAggregateEncoder encoder = getEncoderForAggregateFunction(func.left, func.right);
+			resultBuilder.accumulate(encoder.encodeAggregateLiterals(ctx, ctx.getAggregateFunctionsToRewrite().get(func)));
+		}
+		return resultBuilder.build();
+	}
+
+	private AbstractAggregateEncoder getEncoderForAggregateFunction(AggregateFunctionSymbol function, ComparisonOperator operator) {
+		switch (function) {
+			case COUNT:
+				if (operator == ComparisonOperator.EQ) {
+					return countEqualsEncoder;
+				} else if (operator == ComparisonOperator.LE) {
+					return countLessOrEqualSortingGridEncoder;
+				} else {
+					throw new UnsupportedOperationException("No fitting encoder for aggregate function " + function + "and operator " + operator + "!");
+				}
```

Comment:
tbh, I don't see the benefit in doing that - I still need to check the operator per aggregate function becuase I need to return the right encoder. A possible solution would be to duplicate the aggregate function check within an `if...else if` over operators, but then I'd have to replace the `switch` with ifs and I think it would hurt readability even more.